Ingest-storage alerts (#7702)
grafana/mimir: docs/sources/mimir/manage/mimir-runbooks/_index.md

- Check ingester logs to find details about the error.
- Check Kafka logs and health.

### MimirIngesterFailedToReadRecordsFromKafka

This alert fires when an ingester is failing to read records from Kafka backend.

How it **works**:

> - Ingester connects to Kafka brokers and reads records from it.

Reviewer: ```suggestion
- Ingester connects to Kafka brokers and reads records from it. Records contain write requests committed by distributors.
```

- When ingester fails to read more records from Kafka due to error, ingester logs such error.
- This can be normal if Kafka brokers are restarting, however if read errors continue for some time, alert is raised.

How to **investigate**:

- Check ingester logs to find details about the error.
- Check Kafka logs and health.

### MimirIngesterKafkaFetchErrorsRateTooHigh

This alert fires when an ingester is receiving errors instead of "fetches" from Kafka.

How it **works**:

> - Ingester uses Kafka client to read records from Kafka.

Reviewer: ```suggestion
- Ingester uses Kafka client to read records (containing write requests) from Kafka.
```

- Kafka client can return errors instead of more records.
- If rate of returned errors compared to returned records is too high, alert is raised.
- Kafka client can return errors [documented in the source code](https://github.com/grafana/mimir/blob/24591ae56cd7d6ef24a7cc1541a41405676773f4/vendor/github.com/twmb/franz-go/pkg/kgo/record_and_fetch.go#L332-L366).

How to **investigate**:

- Check ingester logs to find details about the error.
- Check Kafka logs and health.

### MimirStartingIngesterKafkaReceiveDelayIncreasing

This alert fires when "receive delay" reported by ingester during "starting" phase is not decreasing.

How it **works**:

> - When ingester is starting, it needs to fetch and process records from Kafka until preconfigured consumption lag is honored.

Reviewer: ```suggestion
- When ingester is starting, it needs to fetch and process records from Kafka until preconfigured consumption lag is honored. The maximum tolerated lag before an ingester is considered to have caught up reading from a partition at startup can be configured via `-ingest-storage.kafka.max-consumer-lag-at-startup`.
```

> - Each record has a timestamp when it was stored to Kafka. When ingester reads the record, it computes "receive delay" as a difference between current time (when record was read) and time when record was stored to Kafka. This receive delay is reported in metrics.

Reviewer: ```suggestion
- Each record has a timestamp when it was sent to Kafka by the distributor. When ingester reads the record, it computes "receive delay" as a difference between current time (when record was read) and time when record was sent to Kafka. This receive delay is reported in the metric `cortex_ingest_storage_reader_receive_delay_seconds`.
```

Reviewer: We can also mention the dashboard where you can see `cortex_ingest_storage_reader_receive_delay_seconds`. Use the standard format to mention dashboards, so that the auto-link will work in Grafana cloud.

> - Under normal conditions when ingester is processing records faster than records are appearing, receive delay should be decreasing.

Reviewer: ```suggestion
- Under normal conditions when ingester is processing records faster than records are appearing, receive delay should be decreasing, until `-ingest-storage.kafka.max-consumer-lag-at-startup` is honored.
```

- When ingester is starting, and observed "receive delay" is increasing, alert is raised.

How to **investigate**:

> - Check if ingester is fast enough to process all data in Kafka. If not, configure ingesters to start with later offset instead.

Reviewer: > If not, configure ingesters to start with later offset instead.

I'm not super convinced by this. The only solution we currently offer is to start replaying from the partition "end" but it will cause data loss. I think right now we don't have a good strategy to address this, but I would remove "If not, configure ingesters to start with later offset instead" and create an issue to address it instead (e.g. improve parallelism when ingesting data read from Kafka, so we can add more parallelism and eventually vertically scale up ingesters?).

Author: Removed sentence, and created internal issue.

### MimirRunningIngesterReceiveDelayTooHigh

This alert fires when "receive delay" reported by ingester while it's running reaches alert threshold.

How it **works**:

- After ingester start and catches up with records in Kafka, ingester switches to "running" mode.

> - In running mode, ingester continues to process incoming samples from Kafka and continues to report "receive delay". See [`MimirStartingIngesterKafkaReceiveDelayIncreasing`](#MimirStartingIngesterKafkaReceiveDelayIncreasing) runbook for details about this metric.

Reviewer: ```suggestion
- In running mode, ingester continues to process incoming records from Kafka and continues to report "receive delay". See [`MimirStartingIngesterKafkaReceiveDelayIncreasing`](#MimirStartingIngesterKafkaReceiveDelayIncreasing) runbook for details about this metric.
```

> - Under normal conditions when ingester is running and it is processing records faster than records are appearing, receive delay should be stable.

Reviewer: ```suggestion
- Under normal conditions when ingester is running and it is processing records faster than records are appearing, receive delay should be stable and low.
```

- If observed "receive delay" increases and reaches certain threshold, alert is raised.

How to **investigate**:

- Check if ingester is fast enough to process all data in Kafka.

> - If ingesters are too slow, consider scaling ingesters, either vertically (to make ingesters faster), or horizontally to spread incoming series between more ingesters.

Reviewer: > either vertically (to make ingesters faster)

I'm not super convinced that right now it's a viable option, because incoming records will be kept processed sequentially (we don't have a real parallelism there). However, horizontal scaling is an option in this case (but it's not for `MimirRunningIngesterReceiveDelayTooHigh`) because new data will be sharded across more partitions (and ingesters).

Author: Removed mention of vertical scaling. Horizontal scaling also doesn't help immediately, but can help eventually.

### MimirIngesterFailsToProcessRecordsFromKafka

This alert fires when ingester is unable to process incoming records from Kafka due to internal errors. If ingest-storage wasn't used, such push requests would end up with 5xx errors.

How it **works**:

- Ingester reads records from Kafka, and processes them locally. Processing means unmarshalling the data and handling write requests stored in records.

> - Write requests can fail due to "user" or "server" errors. Typical user error is too low limit for number of series. Server error can be for example ingester hitting an instance limit.

Reviewer: > Typical user error 

I would rather say "An example of user error..."

Author: Oh, it's actually "client" error. I will fix it.

- If requests keep failing due to server errors, this alert is raised.

How to **investigate**:

- Check ingester logs to see why requests are failing, and troubleshoot based on that.

### MimirIngesterFailsEnforceStrongConsistencyOnReadPath

This alert fires when too many read-requests with strong consistency are failing.

How it **works**:

- When read request asks for strong-consistency guarantee, ingester will read the last produced offset from Kafka, and wait until record with this offset is consumed.
- If read request times out during this wait, that is considered to be a failure of request with strong-consistency.
- If requests keep failing due to failure to enforce strong-consistency, this alert is raised.

How to **investigate**:

> - Check wait latency of requests with strong-consistency.

Reviewer: Mention the dashboard where you can see it. Use the standard format to mention dashboards, so that the auto-link will work in Grafana cloud.

- Check if ingester needs to process too many records, and whether ingesters need to be scaled up (vertically or horizontally).

> - Consider increasing read-timeout of requests.

Reviewer: Not sure this is a real option. Queries timeout is already pretty high. Increasing timeout will just add more inflight queries in the system. I think we should focus on investigating why the latency is high.

Author: Removed last point.
